### ISSUE TITLE
feature request: change function export format to enable mocking library methods

### ISSUE THREAD

**Issue by bdwain** [rollup/rollup#1682]
I'm not totally sure how difficult/possible this is, but I can look into it. I noticed that if my library exports a function, it usually looks like this after being bundled by rollup.

```js
function foo(){
...
}

export {foo};
```

however, when building this app with webpack (and I assume anything else too because it's a limitation of the es modules spec), code that imports foo from my library can not spy on foo. See [this webpack issue](https://github.com/webpack/webpack/issues/5834) for more details about why it's impossible.

However, rollup seems to be in a position to solve this issue for libraries by instead generating code like this

```js
export function foo(){

}
```
That WOULD be spyable. It seems like every export could be done inline. If not, maybe just functions can?

I think this would be both easier to read and also have the functional benefit of spying enabling. If this can't be done in the core library, I may try to make a plugin to do it. I'm not sure if that's possible yet

**Comment by bdwain** [comment]
I suppose if something is imported wigth a different name than it is exported, you can just say

```js
export const foo = bar;
``

**Comment by bdwain** [comment]
I was able to hack together a script that sort of solves the problems but in a very inelegant way and that required a fork of rollup (to make a 1 line change). I still think a native rollup solution would be better, but this proof of concept at least shows that it would work and be helpful. 

https://gist.github.com/bdwain/47a0f5df3c6eb4712f8bd4aa07955f77

That script could easily be made into a rollup plugin's transformBundle method.

To avoid dealing with name conflicts (which makes the problem a lot more confusing), i changed the getSafeName function in rollup's bundle.js to say `name += '__deconflict';` at the beginning of the function. Now, none of the variables in the file will conflict with the named exports. 

After that, I just ran the script in the linked gist. It removes the named exports that are functions and adds simple exported wrappers for them at the bottom of the file.

```js
export function foo(){return foo__deconflict.apply(undefined, arguments);}
```

This would be not ideal to use in production code because of the extra function wrappers, but for test code it's fine. However, a more complete solution native to rollup would be able to solve this problem more elegantly (without the forced deconflicting), and could avoid the function wrappers completely.

I've been looking through the code trying to understand how it works to see if I could make a PR with this feature. It may take me a while though to learn how everything works since there's a lot going on in the rollup internals. I would appreciate any suggestions anyone has.

**Comment by bdwain** [comment]
hm the deconflicting isn't perfect. Seems to not work in some instances and the original variable name is still being used even though the declaration was changed.

**Comment by lukastaegert** [comment]
Hi,
yes, Rollup's bundling/deconflicting has some issues and yours is not the only one. My plan  is to start re-working this in some way in the near future but I cannot give you a time frame for that. I like the idea  to not extract the exports to the end of the file if at all possible but I think this will be some serious refactoring if we want to do it properly.

I do not like the suggestion of just making variable names always longer as you noted yourself this is neither sure to work nor always necessary. Instead I think the way rollup handles deconflicting should change.

**Comment by lukastaegert** [comment]
I did not know that direct exports have advantages in terms of spyability so thanks for raising this issue!

**Comment by bdwain** [comment]
oh yes i was not suggesting to lengthen the names in actual rollup. That was a hack I came up with to avoid having to deal with renaming variables in my plugin. It didn't even work, but I came up with a different idea that does work and should hold me over until rollup implements this, and it requires no fork of rollup. I basically just require that my library have no conflicts with function export names, and then it's easy to parse the bundle and move the exports around. It works for now lol. https://gist.github.com/bdwain/47a0f5df3c6eb4712f8bd4aa07955f77 if you're interested.

Thanks for looking into it! I really appreciate it! There's not really a good solution out there right now to write a non-tiny library and have the exported functions be spyable, so this will be really helpful. I'm also happy to help in any way I can, but I don't know much about rollup's source yet. 

As far as implementation goes, I feel like the current implementation is close. When doing deconflicting, it just needs to give priority to variables that are actually exported so that they end up with the same name as their export and do not get a $1 suffix or anything like that. I would imagine the same logic that keeps track of that could also be used to add an export statement before them.
